Remove deprecated classes

calebj15 opened the pull request.

mrossinek commented on entanglement_forging/core/wrappers/entanglement_forged_driver.py:
```diff
@@ -48,26 +64,66 @@ def __init__(
-        q_molecule.nuclear_repulsion_energy = self._nuclear_repulsion_energy
+        # Create ParticleNumber property
+        particle_number = ParticleNumber(
+            self._mo_coeff.shape[0], (self._num_alpha, self._num_beta)
+        )
+
```
```suggestion
            self._mo_coeff.shape[0] * 2, (self._num_alpha, self._num_beta)
        )

```
In the `properties` framework we are consistently using the number of spin orbitals (which is twice the number of MOs). Hence you need to multiply this by 2

calebj15 replied:
Done

mrossinek commented on entanglement_forging/core/wrappers/entanglement_forged_driver.py:
```diff
@@ -48,26 +64,66 @@ def __init__(
+            self._mo_coeff.shape[0], (self._num_alpha, self._num_beta)
+        )
+
+        # Convert one body integrals from AO to MO basis
+        one_body_in_mo_basis = np.dot(
+            np.dot(np.transpose(self._mo_coeff), self._hcore), self._mo_coeff
+        )
+
+        # Convert two-body integrals from AO to MO basis
+        dim = self._eri.shape[0]
+        two_body_in_mo_basis = np.zeros((dim, dim, dim, dim))
+        for a_i in range(dim):
+            temp1 = np.einsum("i,i...->...", self._mo_coeff[:, a_i], self._eri)
+            for b_i in range(dim):
+                temp2 = np.einsum("j,j...->...", self._mo_coeff[:, b_i], temp1)
+                temp3 = np.einsum("kc,k...->...c", self._mo_coeff, temp2)
+                two_body_in_mo_basis[a_i, b_i, :, :] = np.einsum(
+                    "ld,l...c->...cd", self._mo_coeff, temp3
+                )
```
You can let the `properties` framework do these transformatiions for you. I'll add suggestions below on how to do that, which will allow you to remove all of this :+1:
See for example how the `PySCFDriver` is doing this:
https://github.com/Qiskit/qiskit-nature/blob/aae9c01b4d02bf5dbaa342c757be2cfef60a31d5/qiskit_nature/drivers/second_quantization/pyscfd/pyscfdriver.py#L620-L639

mrossinek commented on entanglement_forging/core/wrappers/entanglement_forged_driver.py:
```diff
@@ -48,26 +64,66 @@ def __init__(
-        one_body_in_mo_basis = QMolecule.oneeints2mo(self._hcore, self._mo_coeff)
-        q_molecule.mo_onee_ints = one_body_in_mo_basis
-        q_molecule.mo_onee_ints_b = one_body_in_mo_basis
+        # Define the transform from AO to MO
+        ele_basis_xform = ElectronicBasisTransform(
+            ElectronicBasis.AO, ElectronicBasis.MO, self._mo_coeff
+        )
```
Move this above the `ElectronicEnergy` specification

mrossinek commented on entanglement_forging/core/wrappers/entanglement_forged_driver.py:
```diff
@@ -48,26 +64,66 @@ def __init__(
+                )
+
+        # Create ElectronicEnergy property with integrals in both bases and the nuclear repulsion energy
+        integrals: List[ElectronicIntegrals] = []
+        integrals.append(
+            OneBodyElectronicIntegrals(ElectronicBasis.AO, (self._hcore, None))
+        )
+        integrals.append(
+            TwoBodyElectronicIntegrals(
+                ElectronicBasis.AO, (self._eri, None, None, None)
+            )
+        )
+        integrals.append(
+            OneBodyElectronicIntegrals(ElectronicBasis.MO, (one_body_in_mo_basis, None))
+        )
+        integrals.append(
+            TwoBodyElectronicIntegrals(
+                ElectronicBasis.MO,
+                (
+                    two_body_in_mo_basis,
+                    two_body_in_mo_basis,
+                    two_body_in_mo_basis,
+                    None,
+                ),
+            )
+        )
+        electronic_energy = ElectronicEnergy(
+            integrals, nuclear_repulsion_energy=self._nuclear_repulsion_energy
+        )
```
```suggestion
        one_body_ao = OneBodyElectronicIntegrals(ElectronicBasis.AO, (self._hcore, None))
        two_body_ao = TwoBodyElectronicIntegrals(
            ElectronicBasis.AO, (self._eri, None, None, None)
        )

        one_body_mo = one_body_ao.transform_basis(ele_basis_xform)
        two_body_mo = two_body_ao.transform_basis(ele_basis_xform)

        electronic_energy = ElectronicEnergy(
            [one_body_ao, two_body_ao, one_body_mo, two_body_mo],
            nuclear_repulsion_energy=self._nuclear_repulsion_energy,
        )
```
This would be my suggestion for how to let the `ElectronicBasisTransform` do the heavy lifting for you

calebj15 replied:
This is great, much appreciated

mrossinek commented on entanglement_forging/core/wrappers/entanglement_forged_driver.py:
```diff
@@ -48,26 +64,66 @@ def __init__(
-        q_molecule.mo_eri_ints = two_body_in_mo_basis
-        q_molecule.mo_eri_ints_ba = two_body_in_mo_basis
-        q_molecule.mo_eri_ints_bb = two_body_in_mo_basis
+        result = ElectronicStructureDriverResult()
```
I see that you are discarding the `Molecule` information. Is this because you only supplied them to have a working `QMolecule` in the past?
If not, you can follow the procedure here: https://github.com/Qiskit/qiskit-nature/blob/aae9c01b4d02bf5dbaa342c757be2cfef60a31d5/qiskit_nature/drivers/second_quantization/pyscfd/pyscfdriver.py#L528-L539

Note, that the `Molecule` is not a `Property` by itself and thus is set as an attribute of the `result` object.

calebj15 replied:
It looked like the atom_xyz and atom_symbol were left as empty lists, so I wasn't sure if there was ever any molecule geometry loaded into these objects. Am I missing something?

mrossinek replied:
I'm not sure tbh. If you don't really make use of them in this repo, they may well be empty and in that case there is no need for the `Molecule` object to be added to the `ElectronicStructureDriverResult` :+1:

mrossinek reviewed (commented):
Here are some suggestions for the transition to the `properties` framework :+1:

calebj15 commented:
@mrossinek thanks for the review! I think this PR is ready for final review and merge.

In addition to what you looked like, I just removed some Ignis dependencies by calling the new methods (wherever they live). I also removed a section of code that creates a qobj (deprecated), and I just pass the circuits directly to the backend.

mrossinek reviewed (approved):
> In addition to what you looked at, I just removed some Ignis (deprecated) dependencies by calling the new methods (wherever they live). I also removed a section of code that creates a qobj (deprecated), and I just pass the circuits directly to the backend.

I cannot speak for these changes (but they look reasonable enough to me), but in terms of the Nature-related changes this LGTM :+1:
